examples/dark-mode/app/utils/theme-provider.tsx

```diff
@@ -52,43 +85,43 @@ function ThemeProvider({
     return getPreferredTheme();
   });
 
-  const persistTheme = useFetcher();
-  // TODO: remove this when persistTheme is memoized properly
-  const persistThemeRef = useRef(persistTheme);
-  useEffect(() => {
-    persistThemeRef.current = persistTheme;
-  }, [persistTheme]);
-
-  const mountRun = useRef(false);
-
-  useEffect(() => {
-    if (!mountRun.current) {
-      mountRun.current = true;
-      return;
-    }
-    if (!theme) {
-      return;
-    }
-
-    persistThemeRef.current.submit(
-      { theme },
-      { action: "action/set-theme", method: "post" }
-    );
-  }, [theme]);
-
   useEffect(() => {
     const mediaQuery = window.matchMedia(prefersDarkMQ);
     const handleChange = () => {
+      if (specifiedTheme) return;
       setTheme(mediaQuery.matches ? Theme.DARK : Theme.LIGHT);
     };
     mediaQuery.addEventListener("change", handleChange);
     return () => mediaQuery.removeEventListener("change", handleChange);
-  }, []);
+  }, [specifiedTheme]);
 
   return (
-    <ThemeContext.Provider value={[theme, setTheme]}>
-      {children}
-    </ThemeContext.Provider>
+    <ThemeContext.Provider value={theme}>{children}</ThemeContext.Provider>
+  );
+}
+
+function ThemeToggle({
+  children,
+  className,
+}: {
+  children: ReactNode;
+  className?: string;
+}) {
+  const location = useLocation();
+  const searchParams = new URLSearchParams({ redirectTo: location.pathname });
+
+  return (
+    <Form
+      action={`/action/set-theme?${searchParams}`}
+      method="post"
+      className={className}
+    >
+      <Themed
+        dark={<input type="hidden" name="theme" value={Theme.LIGHT} />}
+        light={<input type="hidden" name="theme" value={Theme.DARK} />}
+      />
+      <button type="submit">{children}</button>
+    </Form>
```

Comment:
By using a GET parameter for `redirectTo`, you're opening yourself to an "Open redirection XSS" where a malicious user can pretend to be your website and redirect users elsewhere :

`https://yoursite.com/action/set-theme?redirectTo=https://malwarewebsite.com`

How about making `redirectTo` a hidden input in your form, that would get POSTed instead, defeating the attack?

Reply:
hum 😅, not really an issue since this is handled in your `action`, so it's harder to forge the attack, but still a healthy habit to avoid passing redirects in the path

Reply:
Good point ✅